keps/sig-instrumentation/4785-resource-state-metrics/README.md

```diff
@@ -0,0 +1,1216 @@
+- Once fetched, the controller `unmarshal`s the configuration YAML directly into
+`stores` which are a set of metric `families`, which in turn are a set of 
+`metrics`.
+- Metric `stores` are created based on its respective GVKR (a type that embeds 
+`schema.GroupVersionKind`, `schema.GroupVersionResource` to avoid 
+[plural ambiguities]), and reflectors for the specified resource are
+initialized, and populate the stores on its update.
+- `/metrics` pings on `RSM_MAIN_PORT` trigger the server to write the
+raw metrics, combined with its appropriate header(s), in the response. All
+generated metrics are hardcoded to `gauge`s by design, as Prometheus lacks
+support for some OpenMetrics-specified metrics' types, such as `Info` and
+`StateSets`.
+
+At the moment, the `spec` houses a single `configuration` field, which defines
+the metric generation configuration as follows (please note that the schema is
+fast-moving at this point and may be subject to change based on the [feedback 
+obtained](https://github.com/rexagod/resource-state-metrics/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aopen)):
+
+```yaml
+stores: # Set of metrics stores for each CR we want to generate metrics for.
+  - g: "contoso.com" # CR's group.
+    v: "v1alpha1"    # CR's version.
+    # Both kind and resource names are required to avoid plural ambiguities, see
+    # https://github.com/kubernetes-sigs/kubebuilder/issues/3402.
+    k: "MyPlatform"  # CR's kind.
+    r: "myplatforms" # CR's resource.
+    selectors: # Set of filters to narrow down the selected CRs, may be:
+      field: "metadata.namespace=default" # field selector(s), and (/or),
+      label: "app.kubernetes.io/part-of=sample-controller" # label selector(s).
+    families: # Set of metrics families to generate for the specified CR.
+      - name: "platform_info" # The metric family name, plugged in as-is.
+        help: "Information about a MyPlatform instance" # The help text for the
+                                                        # metric family, plugged
+                                                        # in as-is.
+        metrics: # Set of metrics to generate under the current metrics family.
+          - resolver: "cel" # Preferred resolver to parse the value expressions.
+                            # MAY BE SPECIFIED AT ANY LEVEL.
+                            # INNER RESOLVERS OVERRIDE THE OUTER ONES.
+                            # NO RESOLVER DEFAULTS TO THE "UNSTRUCTURED" ONE.
+            # Set of label-sets to generate for the current metric.
+            labelKeys: # Set of ordered label-keys, static in nature.
+              - "name"
+              - "static_foo"
+            labelValues: # Set of ordered label values, dynamic in nature.
+                         # Therefore, these may contain static values or
+                         # parse-able expressions.
+              - "o.metadata.name" # Parse-able CEL expression.
+              - "static_foo_value" # Static value.
```

Comment:
I feel like having two distinct arrays for label keys and values could be prone to errors. Could we perhaps have them in a map instead? Like:
```yaml
- labels:
  name: "o.metadata.name"
```

Reply:
I see. There's places in the codebase where logic surrounding labels requires ordering, so we'd have to convert `labels` defined at either of the three scopes (`stores`, `families`, and `metrics`) to list first.

Preserving original order can also help us pass results based on the query to the `value` field, which can then work based on that contextual data. We've seen requests like this for CRS which demand a shared context between `labels` and `value`, and an ordering should help users specify that in the `value` field.